File: my-app/frontend/api/bookmark.js
import { Platform } from "react-native";
import { tryMultipleURLs } from "../utils/urlConfig";

// Get the correct base URL based on platform
const getBaseURL = () => {
  if (__DEV__) {
    if (Platform.OS === "ios") {
      // iOS Simulator - use the actual IP address that works
      return "http://192.168.1.73:8080";
    } else if (Platform.OS === "android") {
      // Android emulator needs 10.0.2.2 to access host machine's localhost
      return "http://10.0.2.2:8080";
    }
  }
  // Production URL would go here - updated IP
  return "http://192.168.1.73:8080";
};

// Toggle bookmark (add/remove bookmark)
export const toggleBookmark = async (postId, userId) => {
  try {
    console.log(`🔖 Toggling bookmark for post ${postId}, user ${userId}`);

    const response = await tryMultipleURLs(`/toggle/bookmark/${postId}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        user_id: userId,
      }),
    });

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    const data = await response.json();
    console.log(`✅ Bookmark toggled:`, data);
    return data; // Returns { bookmarked: true/false, message: "..." }
  } catch (error) {
    console.error("❌ Error toggling bookmark:", error);
    throw error;
  }
};

// Get bookmarks for a specific post
export const getBookmarksByPost = async (postId, userId = null) => {
  try {
    console.log(`📚 Fetching bookmarks for post ${postId}, user ${userId}`);

    let url = `/get/bookmarks/${postId}`;
    if (userId) {
      url += `?user_id=${userId}`;
    }

    const response = await tryMultipleURLs(url);

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    const data = await response.json();
    console.log(`✅ Bookmarks fetched:`, data);
    return data; // Returns { bookmark_count: number, user_bookmarked: boolean }
  } catch (error) {
    console.error("❌ Error fetching bookmarks:", error);
    throw error;
  }
};

// Get all bookmarked posts for a user
export const getUserBookmarks = async (userId) => {
  try {
    console.log(`📚 Fetching user bookmarks for user ${userId}`);

    const response = await tryMultipleURLs(`/get/user/bookmarks/${userId}`);

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    const data = await response.json();
    console.log(`✅ User bookmarks fetched:`, data);
    return data; // Returns array of bookmarked posts
  } catch (error) {
    console.error("❌ Error fetching user bookmarks:", error);
    throw error;
  }
};

export default { toggleBookmark, getBookmarksByPost, getUserBookmarks };

File: my-app/backendTwo/handlers/bookmark.go
package handlers

import (
	"encoding/json"
	"net/http"
	"strconv"
	"strings"
	"time"

	"koplaybook.com/backend/models"
)

// ToggleBookmark handles POST /toggle/bookmark/{post_id}
func (h handler) ToggleBookmark(w http.ResponseWriter, r *http.Request) {
	// Add CORS headers
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.Header().Set("Access-Control-Allow-Methods", "POST, OPTIONS")
	w.Header().Set("Access-Control-Allow-Headers", "Content-Type")

	if r.Method == "OPTIONS" {
		w.WriteHeader(http.StatusOK)
		return
	}

	// Extract post ID from URL path /toggle/bookmark/{post_id}
	postIDStr := strings.TrimPrefix(r.URL.Path, "/toggle/bookmark/")
	if postIDStr == "" || postIDStr == r.URL.Path {
		http.Error(w, "Post ID is required", http.StatusBadRequest)
		return
	}

	postID, err := strconv.Atoi(postIDStr)
	if err != nil {
		http.Error(w, "Invalid post ID", http.StatusBadRequest)
		return
	}

	var requestData struct {
		UserID int `json:"user_id"`
	}

	if err := json.NewDecoder(r.Body).Decode(&requestData); err != nil {
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	// Check if the bookmark already exists
	var existingBookmark models.Bookmark
	result := h.DB.Where("post_id = ? AND user_id = ?", postID, requestData.UserID).First(&existingBookmark)

	if result.Error == nil {
		// Bookmark exists, so remove it (unbookmark)
		if err := h.DB.Delete(&existingBookmark).Error; err != nil {
			http.Error(w, "Failed to remove bookmark", http.StatusInternalServerError)
			return
		}

		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]interface{}{
			"message":    "Bookmark removed successfully",
			"bookmarked": false,
		})
	} else {
		// Bookmark doesn't exist, so create a new bookmark
		newBookmark := models.Bookmark{
			PostID:    uint(postID),
			UserID:    uint(requestData.UserID),
			CreatedAt: time.Now(),
		}

		if err := h.DB.Create(&newBookmark).Error; err != nil {
			http.Error(w, "Failed to create bookmark", http.StatusInternalServerError)
			return
		}

		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]interface{}{
			"message":    "Post bookmarked successfully",
			"bookmarked": true,
		})
	}
}

// GetBookmarksByPost handles GET /get/bookmarks/{post_id}?user_id={user_id}
func (h handler) GetBookmarksByPost(w http.ResponseWriter, r *http.Request) {
	// Add CORS headers
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.Header().Set("Access-Control-Allow-Methods", "GET, OPTIONS")
	w.Header().Set("Access-Control-Allow-Headers", "Content-Type")

	if r.Method == "OPTIONS" {
		w.WriteHeader(http.StatusOK)
		return
	}

	// Extract post ID from URL path /get/bookmarks/{post_id}
	postIDStr := strings.TrimPrefix(r.URL.Path, "/get/bookmarks/")
	if postIDStr == "" || postIDStr == r.URL.Path {
		http.Error(w, "Post ID is required", http.StatusBadRequest)
		return
	}

	postID, err := strconv.Atoi(postIDStr)
	if err != nil {
		http.Error(w, "Invalid post ID", http.StatusBadRequest)
		return
	}

	var bookmarks []models.Bookmark
	if err := h.DB.Where("post_id = ?", postID).Find(&bookmarks).Error; err != nil {
		http.Error(w, "Failed to get bookmarks", http.StatusInternalServerError)
		return
	}

	// Get bookmark count
	bookmarkCount := len(bookmarks)

	// Check if current user bookmarked this post (if user_id is provided in query)
	userIDStr := r.URL.Query().Get("user_id")
	var userBookmarked bool = false

	if userIDStr != "" {
		userID, err := strconv.Atoi(userIDStr)
		if err == nil {
			for _, bookmark := range bookmarks {
				if bookmark.UserID == uint(userID) {
					userBookmarked = true
					break
				}
			}
		}
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"post_id":          postID,
		"bookmark_count":   bookmarkCount,
		"user_bookmarked": userBookmarked,
		"bookmarks":       bookmarks,
	})
}

// GetUserBookmarks handles GET /get/user/bookmarks/{user_id}
func (h handler) GetUserBookmarks(w http.ResponseWriter, r *http.Request) {
	// Add CORS headers
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.Header().Set("Access-Control-Allow-Methods", "GET, OPTIONS")
	w.Header().Set("Access-Control-Allow-Headers", "Content-Type")

	if r.Method == "OPTIONS" {
		w.WriteHeader(http.StatusOK)
		return
	}

	// Extract user ID from URL path /get/user/bookmarks/{user_id}
	userIDStr := strings.TrimPrefix(r.URL.Path, "/get/user/bookmarks/")
	if userIDStr == "" || userIDStr == r.URL.Path {
		http.Error(w, "User ID is required", http.StatusBadRequest)
		return
	}

	userID, err := strconv.Atoi(userIDStr)
	if err != nil {
		http.Error(w, "Invalid user ID", http.StatusBadRequest)
		return
	}

	// Get all bookmarks for the user
	var bookmarks []models.Bookmark
	if err := h.DB.Where("user_id = ?", userID).Order("created_at DESC").Find(&bookmarks).Error; err != nil {
		http.Error(w, "Failed to fetch bookmarks", http.StatusInternalServerError)
		return
	}

	// Get the posts for each bookmark
	var bookmarkedPosts []models.Posts
	for _, bookmark := range bookmarks {
		var post models.Posts
		if err := h.DB.First(&post, bookmark.PostID).Error; err == nil {
			bookmarkedPosts = append(bookmarkedPosts, post)
		}
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(bookmarkedPosts)
}
